Support multiple Replaceable and Replacement traits

pchote opened the pull request:
While working to remove the `BuildingInfluence` trait (which will come straight after this) I noticed that the `Replaceable` and `Replacement` traits added by #14862 did not behave well if multiple traits were used on the same actor.

This PR rewrites the logic to address this while also switching `BuildingInfluence` &rarr; `ActorMap`.

A simple testcase is to split the `GAWALL` `Replaceable` defintions over two traits, and to add a second `Replacement` trait to `NAGATE_A` that allows it (but not `NAGATE_B`) to also be placed on GDI walls:

```diff
diff --git a/mods/ts/rules/gdi-support.yaml b/mods/ts/rules/gdi-support.yaml
index 652e0fa75b..d206d4d093 100644
--- a/mods/ts/rules/gdi-support.yaml
+++ b/mods/ts/rules/gdi-support.yaml
@@ -16,8 +16,10 @@ GAWALL:
                CrushClasses: heavywall
        LineBuild:
                NodeTypes: wall, turret
-       Replaceable:
-               Types: GDIGate, GDITower
+       Replaceable@GDIGATE:
+               Types: GDIGate
+       Replaceable@GDITOWER:
+               Types: GDITower
 
 GAGATE_A:
        Inherits: ^Gate_A
diff --git a/mods/ts/rules/nod-support.yaml b/mods/ts/rules/nod-support.yaml
index 1ee8c69286..986e7dd314 100644
--- a/mods/ts/rules/nod-support.yaml
+++ b/mods/ts/rules/nod-support.yaml
@@ -27,8 +27,10 @@ NAGATE_A:
                Prerequisites: nahand, ~structures.nod, ~techlevel.low
        Tooltip:
                Name: Nod Gate
-       Replacement:
+       Replacement@NODGATE:
                ReplaceableTypes: NodGate
+       Replacement@GDITOWER:
+               ReplaceableTypes: GDITower
 
 NAGATE_B:
        Inherits: ^Gate_B
```

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:
```diff
@@ -22,22 +22,52 @@ public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, Bu
+				}
+
+				// The cell contained at least one (not ignored) actor
+				if (checkReplacements && foundActors)
```
If `checkReplacements` is `false`, `foundActors` can never be true either, so this can be simplified to just `foundActors`. (But this is basically just a cosmetic change at this point, I don't really mind keeping it as is if you prefer that for readability.)

pchote committed:
Improve Replaceable logic:

- Remove BuildingInfluence checks
- Support multiple Replaceable/Replacement traits on the same actors
- Fix description typos

pchote commented:
Updated.

phrohdoh reviewed (approved):
LGTM and works in-game with described testcase

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3d6e8ed4ab2b25cec1031bffb6233f0c4b9d7305...c9c4366f85ba04b7c538c310d75a5ceab9b4d5bb)
